### PR TITLE
Adding a new TPP sample for batch normalization

### PR DESCRIPTION
**Pull request by kvoronin-intel**
Summary:

- Refactored existing sample equation/equation_batchnorm.c along the lines of deeplearning/fullyconnected_tpp

<to be extended>

### CODE REVIEW

File: samples/deeplearning/common/dnn_common.h [hunk 1]
```diff
@@ -2429,6 +2458,91 @@ LIBXSMM_INLINE void naive_fusedbatchnorm_fp(naive_fusedbatchnorm_t* param, const
   }
 }
 
+LIBXSMM_INLINE void naive_fusedbatchnorm_fp_fp64(naive_fusedbatchnorm_t* param, const double* input_ptr, double* output_ptr, const double* input_add_ptr,
```
Comment (kvoronin-intel) [inline samples/deeplearning/common/dnn_common.h:2461]: Added naive implementations for fp64 in order to compute reference in fp64 for a more accurate error estimation

File: samples/deeplearning/fused_bn_gn_tpp/Makefile [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+.PHONY: all
+all: $(XFILES)
+
+.PHONY: compile
+compile: $(OBJECTS) $(FTNOBJS)
+
+$(OUTDIR)/layer_example_f32: $(OUTDIR)/.make $(BLDDIR)/layer_example_f32-c.o $(LIBDEP) $(EXTDEP)
+	$(LD) -o $@ $(BLDDIR)/layer_example_f32-c.o $(call cleanld,$(EXTLIB) $(MAINLIB) $(SLDFLAGS) $(LDFLAGS) $(CLDFLAGS))
+
+#$(OUTDIR)/layer_example_bf16_amx: $(OUTDIR)/.make $(BLDDIR)/layer_example_bf16_amx-c.o $(LIBDEP) $(EXTDEP)
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/Makefile:61]: To be removed...

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 3]
```diff
@@ -0,0 +1,1190 @@
...
+  const libxsmm_blasint N  = cfg.N;
+  const libxsmm_blasint CP = cfg.CP;
+  const libxsmm_blasint HW = cfg.HW;
+  const libxsmm_blasint CB = cfg.CB;
+  const libxsmm_blasint num_HW_blocks = cfg.num_HW_blocks;
+
+  /* computing first logical thread */
+  const libxsmm_blasint ltid = my_tid - start_tid;
+
+  /* number of tasks that could be run in parallel for 1d blocking */
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:177]: Only 1d parallelization is implemented (no creation of row- and col-teams) at the moment. Same for bwd.

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 4]
```diff
@@ -0,0 +1,1190 @@
...
+    (work_C / cfg.threads) : ((work_C / cfg.threads) + 1);
+  /* compute thr_begin and thr_end */
+  const libxsmm_blasint thr_begin_C = (ltid * chunksize_C < work_C) ? (ltid * chunksize_C) : work_C;
+  const libxsmm_blasint thr_end_C = ((ltid + 1) * chunksize_C < work_C) ? ((ltid + 1) * chunksize_C) : work_C;
+
+  /* lazy barrier init */
+  libxsmm_barrier_init(cfg.barrier, ltid);
+
+  /* FIXME: Set const modifiers properly? Cannot put const as then
+        reduce_HW_params.in.primary    = &LIBXSMM_VLA_ACCESS(4, inp, n, cp, hwb*(HW/num_HW_blocks), 0, CP, HW, CB); complains */
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:201]: Not sure what's the appropriate way to handle this const-issue. Ideally, const would be good to have here.

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 5]
```diff
@@ -0,0 +1,1187 @@
...
+  libxsmm_matrix_eqn_function  dbeta_func;
+  libxsmm_matrix_eqn_function  din_func;
+  libxsmm_meltwfunction_unary  all_zero_kernel;
+  libxsmm_meltwfunction_binary add_kernel;
+  libxsmm_meltwfunction_unary  copy_kernel;
+  /*my_eltwise_fuse fuse_type;*/
+} my_bn_bwd_config;
+
+my_bn_fwd_config setup_my_bn_fwd(libxsmm_blasint N, libxsmm_blasint CP, libxsmm_blasint HW, libxsmm_blasint CB,
+                                 libxsmm_blasint num_HW_blocks, libxsmm_blasint threads ) {
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:77]: Didn't come up with a better name than the `num_HW_blocks` (taken from the existing sample in equation/)

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 6]
```diff
@@ -0,0 +1,1187 @@
...
+  LIBXSMM_VLA_DECL(3, float, sum_X_X2, ((float*)scratch), CP, CB);  /* [2, CP, CB] */
+  LIBXSMM_ASSUME_ALIGNED(sum_X_X2_, 64);
+  const libxsmm_blasint sum_N_offset = (LIBXSMM_UP2((uintptr_t)(((float*)scratch) + CP * 2 * CB), 64) - ((uintptr_t)(scratch))) / sizeof(float);
+  LIBXSMM_VLA_DECL(3, float, sum_N, ((float*)scratch) + sum_N_offset, N, CB);  /* [CP, N, CB] */
+  LIBXSMM_ASSUME_ALIGNED(sum_N_, 64);
+  const libxsmm_blasint sumsq_N_offset = (LIBXSMM_UP2((uintptr_t)(((float*)scratch) + sum_N_offset + CP * N * CB), 64) - ((uintptr_t)(scratch))) / sizeof(float);
+  LIBXSMM_VLA_DECL(3, float, sumsq_N, ((float*)scratch) + sumsq_N_offset, N, CB);  /* [CP, N, CB] */
+  LIBXSMM_ASSUME_ALIGNED(sumsq_N_, 64);
+
+  { /* stupid block to keep indentation */
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:216]: Will be removed after the code review, leftover from older code

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c (hunk 3 above)
Comment (alheinecke) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:177]: That's fine it's eltwise code.

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c (hunk 5 above)
Comment (alheinecke) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:77]: We should make num_HW_blocks a parameter which is derived inside the setup, depending on size and we should also split HW into one parameter for H and for W. Also CP and CB is blocking/splitting inside of C, so we should provide C and the CB (the inner blocking, I hope this aligns with CNN)

**Review by alheinecke — CHANGES_REQUESTED** [review]
see comments

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c (hunk 5 above)
Comment (kvoronin-intel) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c]: Implemented the changes (also renamed `CB` as `bc` following the naming in fullyconnected_tpp). For now, left num_HW_blocks as internal but hardcoded parameter, without any heuristic.

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 7]
```diff
@@ -0,0 +1,1209 @@
...
+    for ( cpxnt = thr_begin_dN; cpxnt < thr_end_dN; ++cpxnt ) {
+      { /* stupid block to keep indentation */
+        n  = cpxnt%N;
+        cp = cpxnt/N;
+
+        libxsmm_matrix_arg arg_array[8];                                                               /* Private eqn args and params */
+        libxsmm_matrix_eqn_param eqn_param;
+        int hwb, cb;
+
+        /* FIXME: Replace expressions for pgamma, pdgamma etc. with ACCESS? */
```
Comment (kvoronin-intel) [inline samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:650]: Do we want to put LIBXSMM_VLA_ACCESS here? For uniformity, it may make sense but the current form of the expressions in the for-loop looks too good to me to change it.

Comment (alheinecke) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:650]: yes, let's put VLA everywhere

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c (hunk 6 above)
Comment (alheinecke) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:224]: OK

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c [hunk 8]
```diff
@@ -0,0 +1,1209 @@
...
+#define COMPUTE_FP64_REFERENCE
+
+/* include c-based dnn library */
+#ifdef CNN_HEADER
+  #include "dnn_common.h"
+#else /* LIBXSMM sample */
+  #include "../common/dnn_common.h"
+#endif
+
+/* currently unused, as it is not clear that this will be used in CNN
```
[the remark this answers is not in the document]
Comment (alheinecke) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c:30]: you mean my_pass? It's for fullyconnected in BWD as we have two BWD passes (BWD_D and BWD_W) and we can run both, or just one of the them. For BN/ GN we only have BWD_D. Since we have FWD and BWD calls we don't need that here.

**Comment by alheinecke** [comment]
should we merge in this state or are some more commits comming?

**Comment by alheinecke** [comment]
BTW, while it's fine to skip padding/stride options. We need to support:

    if ( norm_type == 0 ) {
      if ( fuse_type == 0 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN;
      } else if ( fuse_type == 1 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN_RELU;
      } else if ( fuse_type == 2 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN_ELTWISE;
      } else if ( fuse_type == 3 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN_ELTWISE_RELU;
      } else if ( fuse_type == 4 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN_RELU_WITH_MASK;
      } else if ( fuse_type == 5 ) {
        fusedbatchnorm_desc.fuse_ops = LIBXSMM_DNN_FUSEDBN_OPS_BN_ELTWISE_RELU_WITH_MASK;
      } else {
        /* shouldn't happen */
        return -1;
      }
    } 

as in the original driver.

**Comment by kvoronin-intel** [comment]
Tue main goal of opening this PR was to gain general feedback. I think this is done so now I'm changing the state to WIP. Let's keep this PR pending (I will update it following the comments from @alheinecke and with changes implied by integration for CNN, which is in progress right now)

**Comment by alheinecke** [comment]
changed the base to dnn_refactor
please split the code into driver and implementation, implementation goes into samples/op_lib_tpp

**Comment by alheinecke** [comment]
you might want to merge dnn_refactor into your branch such that all the weird conflicts go away

**Review by alheinecke — COMMENTED** [review]
while the batchnorm stuff is developing well, why are these 2/5 cnn files part of the PR?

File: samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c (hunk 7 above)
Comment (kvoronin-intel) [reply samples/deeplearning/fused_bn_gn_tpp/layer_example_f32.c]: Done

File: samples/deeplearning/fusedbn_tpp/run_test.sh [hunk 9]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+set -eo pipefail
+
+# Source the compiler
+
+#export OMP_NUM_THREADS=1
+export CHECK=1
+
+iters=10
```
Comment (alheinecke) [inline samples/deeplearning/fusedbn_tpp/run_test.sh:9]: Can please add layers, e.g. copy from:
https://github.com/libxsmm/libxsmm/blob/master/samples/deeplearning/fusedbndriver/run_resnet50.sh#L99

**Review by alheinecke — CHANGES_REQUESTED** [review]
very minor thing in the test script:
please copy over all resnet test cases. 

the code itself is ready for merge

File: samples/deeplearning/fusedbn_tpp/run_test.sh (hunk 9 above)
Comment (kvoronin-intel) [reply samples/deeplearning/fusedbn_tpp/run_test.sh:9]: Is it still fine that the code doesn't support non-unit strides and non-zero padding (these cases are present in the resnet layers in the script)?

Comment (alheinecke) [reply samples/deeplearning/fusedbn_tpp/run_test.sh:9]: yes, pytroch doesn't support physically padded tensor easily and the stride was to fuse subsampling, but we do this in pytorch in the convolutions,
